sdk/azidentity/chained_token_credential.go

```diff
@@ -50,35 +54,72 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 	if options == nil {
 		options = &ChainedTokenCredentialOptions{}
 	}
-	return &ChainedTokenCredential{sources: cp, name: "ChainedTokenCredential", retrySources: options.RetrySources}, nil
+	mut := sync.RWMutex{}
+	return &ChainedTokenCredential{
+		cond:         sync.NewCond(&mut),
+		mut:          &mut,
+		sources:      cp,
+		name:         "ChainedTokenCredential",
+		retrySources: options.RetrySources,
+	}, nil
 }
 
 // GetToken calls GetToken on the chained credentials in turn, stopping when one returns a token. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	if c.successfulCredential != nil && !c.retrySources {
-		return c.successfulCredential.GetToken(ctx, opts)
+	if !c.retrySources {
```

Comment:
Before I dig into this, is it possible to reuse the [`ExpiringResource`](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/internal/shared/expiring_resource.go) for this?  Ideally, we'd move this machinery into the `internal` module so it can be used in `azcore` and here.

Reply:
There's probably a way to make it work but it would be awkward. The value here never expires and exclusive locking around it is an unnecessary performance hit.

Reply:
If it's going to make it more complicated, then it's not worth it.  The pattern looked similar, so I was just curious if this possibility had been looked at.